[Bug]: MCP SSE endpoint not working

### crawl4ai version

0.8.5

### Expected Behavior

MCP SSE works as normal, can be visited via web browser.
Body should look like

```
event: endpoint
data: /mcp/messages/?session_id=756c37c8063b488b8494d8f9e0ef9efa
```

### Current Behavior

Visiting the endpoint via web browser or any MCP tool usage produces an HTTP 500 Internal Server Error.

### Is this reproducible?

Yes

### Inputs Causing the Bug

```bash
Just navigating to http://localhost:11235/mpc/sse triggers the error
```

### Steps to Reproduce

```bash
1. Run crawl4ai
2. Visit http://localhost:11235/mpc/sse in a web browser or MCP capable client
3. Observe the failure.
```

### Code snippets

```python
Docker compose used to run crawl4AI:

x-base-config: &base-config
  ports:
    - "11235:11235"
  volumes:
    - /dev/shm:/dev/shm  # Chromium performance
  deploy:
    resources:
      limits:
        memory: 4G
      reservations:
        memory: 1G
  restart: unless-stopped
  healthcheck:
    test: ["CMD", "curl", "-f", "http://localhost:11235/health"]
    interval: 30s
    timeout: 10s
    retries: 3
    start_period: 40s
  user: "appuser"

services:
  crawl4ai:
    # 1. Default: Pull multi-platform test image from Docker Hub
    # 2. Override with local image via: IMAGE=local-test docker compose up
    image: unclecode/crawl4ai:0.8.5

    # Inherit shared config
    <<: *base-config


```

### OS

Linux running crawl4ai inside docker.

### Python version

the one used in the docker container for 0.8.5

### Browser

_No response_

### Browser version

_No response_

### Error logs & Screenshots (if applicable)

```
crawl4ai-1  | [2026-03-23 00:25:09 +0000] [15] [ERROR] Exception in ASGI application
crawl4ai-1  | Traceback (most recent call last):
crawl4ai-1  |   File "/usr/local/lib/python3.12/site-packages/uvicorn/protocols/http/h11_impl.py", line 410, in run_asgi
crawl4ai-1  |     result = await app(  # type: ignore[func-returns-value]
crawl4ai-1  |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
crawl4ai-1  |   File "/usr/local/lib/python3.12/site-packages/uvicorn/middleware/proxy_headers.py", line 60, in __call__
crawl4ai-1  |     return await self.app(scope, receive, send)
crawl4ai-1  |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
crawl4ai-1  |   File "/usr/local/lib/python3.12/site-packages/fastapi/applications.py", line 1160, in __call__
crawl4ai-1  |     await super().__call__(scope, receive, send)
crawl4ai-1  |   File "/usr/local/lib/python3.12/site-packages/starlette/applications.py", line 107, in __call__
crawl4ai-1  |     await self.middleware_stack(scope, receive, send)
crawl4ai-1  |   File "/usr/local/lib/python3.12/site-packages/starlette/middleware/errors.py", line 186, in __call__
crawl4ai-1  |     raise exc
crawl4ai-1  |   File "/usr/local/lib/python3.12/site-packages/starlette/middleware/errors.py", line 164, in __call__
crawl4ai-1  |     await self.app(scope, receive, _send)
crawl4ai-1  |   File "/usr/local/lib/python3.12/site-packages/starlette/middleware/base.py", line 191, in __call__
crawl4ai-1  |     with recv_stream, send_stream, collapse_excgroups():
crawl4ai-1  |                                    ^^^^^^^^^^^^^^^^^^^^
crawl4ai-1  |   File "/usr/local/lib/python3.12/contextlib.py", line 158, in __exit__
crawl4ai-1  |     self.gen.throw(value)
crawl4ai-1  |   File "/usr/local/lib/python3.12/site-packages/starlette/_utils.py", line 87, in collapse_excgroups
crawl4ai-1  |     raise exc
crawl4ai-1  |   File "/usr/local/lib/python3.12/site-packages/starlette/middleware/base.py", line 193, in __call__
crawl4ai-1  |     response = await self.dispatch_func(request, call_next)
crawl4ai-1  |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
crawl4ai-1  |   File "/app/server.py", line 250, in add_security_headers
crawl4ai-1  |     resp = await call_next(request)
crawl4ai-1  |            ^^^^^^^^^^^^^^^^^^^^^^^^
crawl4ai-1  |   File "/usr/local/lib/python3.12/site-packages/starlette/middleware/base.py", line 168, in call_next
crawl4ai-1  |     raise app_exc from app_exc.__cause__ or app_exc.__context__
crawl4ai-1  |   File "/usr/local/lib/python3.12/site-packages/starlette/middleware/base.py", line 144, in coro
crawl4ai-1  |     await self.app(scope, receive_or_disconnect, send_no_error)
crawl4ai-1  |   File "/usr/local/lib/python3.12/site-packages/prometheus_fastapi_instrumentator/middleware.py", line 177, in __call__
crawl4ai-1  |     raise exc
crawl4ai-1  |   File "/usr/local/lib/python3.12/site-packages/prometheus_fastapi_instrumentator/middleware.py", line 175, in __call__
crawl4ai-1  |     await self.app(scope, receive, send_wrapper)
crawl4ai-1  |   File "/usr/local/lib/python3.12/site-packages/starlette/middleware/exceptions.py", line 63, in __call__
crawl4ai-1  |     await wrap_app_handling_exceptions(self.app, conn)(scope, receive, send)
crawl4ai-1  |   File "/usr/local/lib/python3.12/site-packages/starlette/_exception_handler.py", line 53, in wrapped_app
crawl4ai-1  |     raise exc
crawl4ai-1  |   File "/usr/local/lib/python3.12/site-packages/starlette/_exception_handler.py", line 42, in wrapped_app
crawl4ai-1  |     await app(scope, receive, sender)
crawl4ai-1  |   File "/usr/local/lib/python3.12/site-packages/fastapi/middleware/asyncexitstack.py", line 18, in __call__
crawl4ai-1  |     await self.app(scope, receive, send)
crawl4ai-1  |   File "/usr/local/lib/python3.12/site-packages/starlette/routing.py", line 716, in __call__
crawl4ai-1  |     await self.middleware_stack(scope, receive, send)
crawl4ai-1  |   File "/usr/local/lib/python3.12/site-packages/starlette/routing.py", line 736, in app
crawl4ai-1  |     await route.handle(scope, receive, send)
crawl4ai-1  |   File "/usr/local/lib/python3.12/site-packages/starlette/routing.py", line 290, in handle
crawl4ai-1  |     await self.app(scope, receive, send)
crawl4ai-1  |   File "/usr/local/lib/python3.12/site-packages/starlette/routing.py", line 78, in app
crawl4ai-1  |     await wrap_app_handling_exceptions(app, request)(scope, receive, send)
crawl4ai-1  |   File "/usr/local/lib/python3.12/site-packages/starlette/_exception_handler.py", line 53, in wrapped_app
crawl4ai-1  |     raise exc
crawl4ai-1  |   File "/usr/local/lib/python3.12/site-packages/starlette/_exception_handler.py", line 42, in wrapped_app
crawl4ai-1  |     await app(scope, receive, sender)
crawl4ai-1  |   File "/usr/local/lib/python3.12/site-packages/starlette/routing.py", line 75, in app
crawl4ai-1  |     response = await f(request)
crawl4ai-1  |                      ^^^^^^^^^^
crawl4ai-1  | TypeError: attach_mcp.<locals>._mcp_sse_handler() missing 2 required positional arguments: 'receive' and 'send'
```

--- a/deploy/docker/mcp_bridge.py
+++ b/deploy/docker/mcp_bridge.py
@@ -223,13 +223,16 @@ def attach_mcp(
     # ── SSE transport (raw ASGI — avoids Starlette middleware conflict) ──
     sse = SseServerTransport(f"{base}/messages/")
 
-    async def _mcp_sse_handler(scope, receive, send):
-        """Raw ASGI handler for SSE — the MCP SDK calls (scope, receive, send)
-        internally, so wrapping with @app.get() causes an AssertionError (#1594)."""
-        async with sse.connect_sse(scope, receive, send) as (read_stream, write_stream):
-            await mcp.run(read_stream, write_stream, init_opts)
+    # Starlette's Route wraps plain async functions in request_response(),
+    # which calls handler(request) instead of handler(scope, receive, send).
+    # Using a callable class bypasses this — Route passes classes through
+    # as raw ASGI apps.  See #1594, #1850.
+    class _MCPSseApp:
+        async def __call__(self, scope, receive, send):
+            async with sse.connect_sse(scope, receive, send) as (read_stream, write_stream):
+                await mcp.run(read_stream, write_stream, init_opts)
 
-    app.routes.append(Route(f"{base}/sse", endpoint=_mcp_sse_handler))
+    app.routes.append(Route(f"{base}/sse", endpoint=_MCPSseApp()))
     app.routes.append(Mount(f"{base}/messages", app=sse.handle_post_message))
 
     # ── schema endpoint ───────────────────────────────────────

--- a/tests/test_issue_1850_mcp_sse.py
+++ b/tests/test_issue_1850_mcp_sse.py
@@ -1,0 +1,172 @@
+"""
+Tests for issue #1850: MCP SSE endpoint not working.
+
+Starlette's Route wraps async functions in request_response(), which calls
+handler(request) instead of handler(scope, receive, send). The fix uses a
+callable class to bypass this wrapping.
+
+These tests verify the ASGI routing behavior without requiring a running
+MCP server or Docker.
+"""
+
+import inspect
+import pytest
+from starlette.routing import Route
+from starlette.testclient import TestClient
+from starlette.applications import Starlette
+from starlette.responses import PlainTextResponse
+
+
+# -- Core issue: Route wrapping behavior --
+
+class TestRouteWrappingBehavior:
+    """Verify that Starlette Route wraps functions but not class instances."""
+
+    def test_async_function_is_wrapped(self):
+        """An async function endpoint gets wrapped in request_response()."""
+        async def handler(scope, receive, send):
+            pass
+
+        r = Route("/test", endpoint=handler)
+        # Route wraps it — r.app is NOT handler
+        assert r.app is not handler
+
+    def test_callable_class_is_not_wrapped(self):
+        """A callable class instance is treated as raw ASGI (not wrapped)."""
+        class Handler:
+            async def __call__(self, scope, receive, send):
+                pass
+
+        h = Handler()
+        r = Route("/test", endpoint=h)
+        # Route passes it through — r.app IS handler
+        assert r.app is h
+
+    def test_async_function_is_function(self):
+        """Confirm async def is detected as function by inspect."""
+        async def handler(scope, receive, send):
+            pass
+        assert inspect.isfunction(handler)
+
+    def test_callable_class_is_not_function(self):
+        """Confirm callable class is NOT detected as function by inspect."""
+        class Handler:
+            async def __call__(self, scope, receive, send):
+                pass
+        assert not inspect.isfunction(Handler())
+
+
+# -- ASGI handler receives correct arguments --
+
+class TestASGIHandlerArgs:
+    """Verify that the callable class receives scope/receive/send correctly."""
+
+    def test_callable_class_receives_asgi_args(self):
+        """A callable class mounted via Route should get scope, receive, send."""
+        received_args = {}
+
+        class Handler:
+            async def __call__(self, scope, receive, send):
+                received_args["scope_type"] = scope["type"]
+                response = PlainTextResponse("ok")
+                await response(scope, receive, send)
+
+        app = Starlette(routes=[Route("/test", endpoint=Handler())])
+        client = TestClient(app)
+        resp = client.get("/test")
+        assert resp.status_code == 200
+        assert received_args["scope_type"] == "http"
+
+    def test_async_function_receives_request_not_asgi(self):
+        """An async function mounted via Route gets Request, not raw ASGI."""
+        received_type = {}
+
+        async def handler(*args, **kwargs):
+            received_type["arg_count"] = len(args)
+            if args and hasattr(args[0], "url"):
+                received_type["is_request"] = True
+                return PlainTextResponse("ok")
+            received_type["is_request"] = False
+
+        app = Starlette(routes=[Route("/test", endpoint=handler)])
+        client = TestClient(app)
+        resp = client.get("/test")
+        # Starlette wraps it and passes Request object (1 arg)
+        assert received_type.get("is_request") is True
+
+
+# -- MCP bridge SSE handler structure --
+
+class TestMCPBridgeSSEHandler:
+    """Verify the mcp_bridge SSE handler is correctly structured."""
+
+    def test_mcp_bridge_uses_callable_class(self):
+        """The SSE handler in mcp_bridge should be a callable class, not a function."""
+        # Import and check the source
+        import importlib.util
+        spec = importlib.util.spec_from_file_location(
+            "mcp_bridge_check",
+            "deploy/docker/mcp_bridge.py"
+        )
+        # We can't fully import mcp_bridge (needs Docker deps), so check source
+        with open("deploy/docker/mcp_bridge.py") as f:
+            source = f.read()
+
+        # Should have a class-based handler
+        assert "class _MCPSseApp" in source
+        assert "async def __call__(self, scope, receive, send)" in source
+
+    def test_mcp_bridge_no_async_def_sse_handler(self):
+        """Should NOT have a plain async def _mcp_sse_handler."""
+        with open("deploy/docker/mcp_bridge.py") as f:
+            source = f.read()
+
+        # The old buggy pattern should be gone
+        assert "async def _mcp_sse_handler(scope, receive, send)" not in source
+
+    def test_mcp_bridge_route_uses_class_instance(self):
+        """Route should be created with _MCPSseApp() instance, not a function."""
+        with open("deploy/docker/mcp_bridge.py") as f:
+            source = f.read()
+
+        assert "_MCPSseApp()" in source
+
+
+# -- Regression: ensure Route + callable class pattern works end-to-end --
+
+class TestRouteCallableClassEndToEnd:
+    """End-to-end test that a callable class works as a Route endpoint."""
+
+    def test_sse_like_handler(self):
+        """Simulate an SSE-like raw ASGI handler via Route."""
+        class SSEHandler:
+            async def __call__(self, scope, receive, send):
+                response = PlainTextResponse(
+                    "event: endpoint\ndata: /test\n\n",
+                    media_type="text/event-stream",
+                )
+                await response(scope, receive, send)
+
+        app = Starlette(routes=[Route("/sse", endpoint=SSEHandler())])
+        client = TestClient(app)
+        resp = client.get("/sse")
+        assert resp.status_code == 200
+        assert "event: endpoint" in resp.text
+
+    def test_multiple_routes_with_mixed_handlers(self):
+        """Callable class and regular function handlers can coexist."""
+        class RawHandler:
+            async def __call__(self, scope, receive, send):
+                response = PlainTextResponse("raw")
+                await response(scope, receive, send)
+
+        async def regular_handler(request):
+            return PlainTextResponse("regular")
+
+        app = Starlette(routes=[
+            Route("/raw", endpoint=RawHandler()),
+            Route("/regular", endpoint=regular_handler),
+        ])
+        client = TestClient(app)
+        assert client.get("/raw").text == "raw"
+        assert client.get("/regular").text == "regular"

